test/rekt/apiserversource_test.go

```diff
@@ -0,0 +1,181 @@
+func TestApiServerSourceValidationWebhookConfigurationOnCreate(t *testing.T) {
+	t.Parallel()
+
+	srcname := feature.MakeRandomK8sName("apiserversource")
+
+	ctx, env := global.Environment()
+	t.Cleanup(env.Finish)
+
+	f := feature.NewFeatureNamed("ApiServerSource webhook is configured correctly.")
+
+	f.Stable("ApiServerSource webhook").
+		Must("reject invalid spec on resource creation", createApiServerSourceWithInvalidSpec(srcname))
+
+	env.Test(ctx, t, f)
+}
+
+// TestApiServerSourceValidationWebhookConfigurationOnUpdate tests if the webhook
+// is configured correctly for ApiServerSource validation on resource update.
+func TestApiServerSourceValidationWebhookConfigurationOnUpdate(t *testing.T) {
+	t.Parallel()
+
+	srcname := feature.MakeRandomK8sName("apiserversource")
+
+	ctx, env := global.Environment()
+	t.Cleanup(env.Finish)
+
+	f := feature.NewFeatureNamed("ApiServerSource webhook is configured correctly.")
+
+	f.Setup("Set ApiServerSource Name", setApiServerSourceName(srcname))
+
+	f.Setup("Create valid ApiServerSource", createApiServerSourceWithValidSpec(srcname))
+
+	f.Stable("ApiServerSource webhook").
+		Must("reject invalid spec", updateApiServerSourceWithInvalidSpec())
+
+	env.Test(ctx, t, f)
+}
+
+func createApiServerSourceWithValidSpec(name string) func(ctx context.Context, t feature.T) {
+	return func(ctx context.Context, t feature.T) {
+		_, err := apiserversource.InstallLocalYaml(ctx, name, apiServerSourceWithValidSpec())
+
+		// we don't care if the resource gets ready or not as we only concerned about the webhook
+
+		if err != nil {
+			t.Error("ApiServerResource with valid spec cannot be applied", err)
+		}
+	}
+}
+
+func createApiServerSourceWithInvalidSpec(name string) func(ctx context.Context, t feature.T) {
+	return func(ctx context.Context, t feature.T) {
+		// generate a valid spec, then change the event mode with an invalid value, then
+		// try actually applying the resource
+		_, err := apiserversource.InstallLocalYaml(ctx, name,
+			apiServerSourceWithValidSpec(), apiserversource.WithEventMode("Unknown"))
+
+		if err != nil {
+			// all good, error is expected
+			assert.EqualError(t, err, `admission webhook "validation.webhook.eventing.knative.dev" denied the request: validation failed: invalid value: Unknown: spec.mode`)
+		} else {
+			t.Error("expected ApiServerResource to reject invalid spec.")
+		}
+	}
+}
+
+func updateApiServerSourceWithInvalidSpec() func(ctx context.Context, t feature.T) {
+	return func(ctx context.Context, t feature.T) {
+		apiServerSource := getApiServerSource(ctx, t)
+
+		apiServerSource.Spec.EventMode = "Unknown"
+
+		_, err := Client(ctx).ApiServerSources.Update(ctx, apiServerSource, metav1.UpdateOptions{})
+
+		if err != nil {
+			// all good, error is expected
+			assert.EqualError(t, err, `admission webhook "validation.webhook.eventing.knative.dev" denied the request: validation failed: invalid value: Unknown: spec.mode`)
```

Comment:
Can't test the "update" case this way. I was thinking
- I could just create something an ApiServerSource with valid spec 
- The sink doesn't have to exist as I don't care if the source is ready or not
- Update its spec to become invalid
- Apply the invalid spec
- Webhook returns an error

This doesn't work because, the reconciler updates the resource too often and there is a race between the test fetching&updating AND the reconciler. I get "the object has been modified; please apply your changes to the latest version and try again".

My focus for this PR was to make sure the webhook is configured correctly. However, I don't think I can get around w/o making sure the ApiServerSource is ready and the reconciler stops updating the resource.

I need to do some modifications.

Reply:
Ok, modifications done